### PR TITLE
Fix reconciliation errors if a referenced secret is deleted

### LINKED ISSUE #216: Reconciliation errors if a referenced secret is deleted.
### What happened?

If a secret is deleted before a terrajet resource is deleted, the following warning will be displayed and the resource will get stuck: 

```Warning  CannotConnectToProvider  23s (x334 over 6m11s)  managed/postgresql.azure.jet.crossplane.io/v1alpha1, kind=server  cannot get a terraform workspace for resource: cannot create a new file producer: cannot get sensitive parameters: cannot get secret value for &{{myazurepgserver-admin-password all-in-one-selfservice} password}: cannot get secret data: Secret "myazurepgserver-admin-password" not found```

### How can we reproduce it?

- Create resource with a secret
- Delete referenced secret before deleting the resource

### PR DESCRIPTION
**Pull request by ezgidemirel**
Signed-off-by: ezgidemirel <ezgidemirel91@gmail.com>

### Description of your changes

Fixes #216 

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

I replaced terrajet version consumed in `provider-jet-azure` to this branch in my local environment and created a PostgreSqlDB resource. Once it's ready, I first deleted the k8s secret referenced by `administratorLoginPasswordSecretRef` field and then deleted the MR. 

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: pkg/resource/sensitive.go [hunk 1]
```diff
@@ -241,7 +226,7 @@ func GetSensitiveParameters(ctx context.Context, client SecretClient, from runti
 					return errors.Wrapf(err, errFmtCannotGetSecretKeySelector, expandedJSONPath)
 				}
 				sensitive, err = client.GetSecretValue(ctx, *sel)
-				if err != nil {
+				if err != nil && !kerrors.IsNotFound(err) {
```
Comment (turkenh) [inline pkg/resource/sensitive.go:229]: We can use [resource.IgnoreNotFound(err)](https://github.com/crossplane/crossplane-runtime/blob/428b7c3903756bb0dcf5330f40298e1fa0c34301/pkg/resource/resource.go#L207) instead.

File: pkg/resource/sensitive.go [hunk 2]
```diff
@@ -203,22 +202,6 @@ func GetSensitiveParameters(ctx context.Context, client SecretClient, from runti
 			}
 
 			switch v.(type) {
-			case map[string]interface{}:
```
Comment (turkenh) [inline pkg/resource/sensitive.go:206]: What would the behavior be in the case of `map[string]secretRef` added to generated schema but we are dropping support here?

File: pkg/resource/sensitive.go [hunk 3]
```diff
@@ -227,7 +210,9 @@ func GetSensitiveParameters(ctx context.Context, client SecretClient, from runti
 				var sensitives []interface{}
 				for _, s := range *sel {
 					sensitive, err = client.GetSecretValue(ctx, s)
-					if err != nil {
+					if err != nil && kerrors.IsNotFound(err) {
+						sensitive = []byte("")
```
Comment (turkenh) [inline pkg/resource/sensitive.go:214]: So, we have tested/validated terraform does not complain about having sensitive as an empty secret? e.g. password validation

File: pkg/resource/sensitive.go (hunk 1 above)
Comment (ezgidemirel) [reply pkg/resource/sensitive.go]: done

File: pkg/resource/sensitive.go (hunk 2 above)
Comment (ezgidemirel) [reply pkg/resource/sensitive.go]: We were seeing following error message on the resource:
```
Warning  CannotConnectToProvider  9s (x25 over 23s)  managed/dbforpostgresql.azure.jet.crossplane.io/v1alpha2, kind=server  cannot get a terraform workspace for resource: cannot create a new file producer: cannot get sensitive parameters: cannot unmarshal value from JSON: json: cannot unmarshal string into Go value of type v1.SecretKeySelector
```
Thanks to @sergenyalcin , the issue is fixed now.

File: pkg/resource/sensitive.go (hunk 3 above)
Comment (ezgidemirel) [reply pkg/resource/sensitive.go:248]: That's correct. I have tested the behavior with both terraform cli (after updating the state and main.tf files) and with provider-jet-azure after consuming my change.

File: pkg/resource/sensitive.go [hunk 4]
```diff
@@ -212,7 +213,9 @@ func GetSensitiveParameters(ctx context.Context, client SecretClient, from runti
 					sensitives := make(map[string]interface{})
 					for key, value := range *sel {
 						sensitive, err = client.GetSecretValue(ctx, value)
-						if err != nil {
+						if err != nil && kerrors.IsNotFound(err) {
+							sensitive = []byte("")
+						} else if err != nil {
 							return errors.Wrapf(err, errFmtCannotGetSecretValue, sel)
 						}
```
Comment (turkenh) [inline pkg/resource/sensitive.go:220]: nitpick:

```suggestion
						if resource.IgnoreNotFound(err) != nil {
							return errors.Wrapf(err, errFmtCannotGetSecretValue, sel)
						} 
						if kerrors.IsNotFound(err) {
							sensitive = []byte("")
						}
```

File: pkg/resource/sensitive.go [hunk 5]
```diff
@@ -241,7 +244,9 @@ func GetSensitiveParameters(ctx context.Context, client SecretClient, from runti
 				var sensitives []interface{}
 				for _, s := range *sel {
 					sensitive, err = client.GetSecretValue(ctx, s)
-					if err != nil {
+					if err != nil && kerrors.IsNotFound(err) {
```
Comment (turkenh) [inline pkg/resource/sensitive.go:247]: same nitpick as above for better readability.

File: pkg/resource/sensitive.go [hunk 6]
```diff
@@ -212,7 +213,9 @@ func GetSensitiveParameters(ctx context.Context, client SecretClient, from runti
 					sensitives := make(map[string]interface{})
 					for key, value := range *sel {
 						sensitive, err = client.GetSecretValue(ctx, value)
-						if err != nil {
+						if err != nil && kerrors.IsNotFound(err) {
+							sensitive = []byte("")
```
Comment (turkenh) [inline pkg/resource/sensitive.go:217]: A note on why we need this would be helpful for further readers.

**Review by turkenh — APPROVED** [review]
Added some nitpicks, looking good otherwise!
Thank you @ezgidemirel!

File: pkg/resource/sensitive.go (hunk 4 above)
Comment (ezgidemirel) [reply pkg/resource/sensitive.go]: done

File: pkg/resource/sensitive.go (hunk 6 above)
Comment (ezgidemirel) [reply pkg/resource/sensitive.go]: done

File: pkg/resource/sensitive.go (hunk 5 above)
Comment (ezgidemirel) [reply pkg/resource/sensitive.go]: done

**Pull request merged**